### PR TITLE
Typed ast allocator

### PR DESCRIPTION
**Pull request by lushoBarlett**
This is the sequel of the last one I did, where we replace `unique_ptr`s for a `NodeAllocator<...>`.

### CODE REVIEW

File: src/ct_eval.cpp [hunk 1]
```diff
@@ -11,89 +11,98 @@ namespace TypeChecker {
...
+    TypedAST::FunctionLiteral* ast, TypeChecker& tc, TypedAST::Allocator& alloc) {
 
 	// TODO: type hints
 
 	assert(ast->m_body->type() == TypedASTTag::Block);
-	auto body = static_cast<TypedAST::Block*>(ast->m_body.get());
-	for (auto&& child : body->m_body)
-		child = ct_eval(std::move(child), tc);
+	auto body = static_cast<TypedAST::Block*>(ast->m_body);
+	for (auto child : body->m_body)
```
Comment (SebastianMestre) [inline src/ct_eval.cpp:30]: iterate by ref

File: src/typed_ast.hpp [hunk 2]
```diff
@@ -166,90 +168,90 @@ struct Identifier : public TypedAST {
...
+	TypedAST* m_body;
 
 	WhileStatement()
 	    : TypedAST {TypedASTTag::WhileStatement} {}
 };
 
 struct StructExpression : public TypedAST {
-	std::vector<Identifier> m_fields;
-	std::vector<Own<TypedAST>> m_types;
+	std::vector<Identifier*> m_fields;
```
Comment (SebastianMestre) [inline src/typed_ast.hpp:245]: why are these pointers now?

File: src/typed_ast.hpp [hunk 3]
```diff
@@ -60,7 +62,7 @@ struct Declaration : public TypedAST {
 
 // doesnt have a ast_vtype
 struct DeclarationList : public TypedAST {
-	std::vector<Own<Declaration>> m_declarations;
+	std::vector<Declaration*> m_declarations;
```
Comment (SebastianMestre) [inline src/typed_ast.hpp:65]: I feel like this indirection is not necessary, but I'm not gonna bother you with that now.

File: src/typed_ast.hpp [hunk 4]
```diff
@@ -128,23 +130,23 @@ struct ObjectLiteral : public TypedAST {
...
 	DictionaryLiteral()
 	    : TypedAST {TypedASTTag::DictionaryLiteral} {}
 };
 
 struct FunctionLiteral : public TypedAST {
 	MonoId m_return_type;
-	Own<TypedAST> m_body;
-	std::vector<Declaration> m_args;
+	TypedAST* m_body;
+	std::vector<Declaration*> m_args;
```
Comment (SebastianMestre) [inline src/typed_ast.hpp:149]: Why are these pointers now?

File: src/typed_ast.cpp [hunk 5]
```diff
@@ -4,222 +4,225 @@
...
 	for (auto arg : ast->m_args) {
 		assert(arg->type() == ASTTag::Declaration);
-		auto* decl = static_cast<AST::Declaration*>(arg);
+		auto decl = static_cast<AST::Declaration*>(arg);
 
-		Declaration typed_decl;
-		typed_decl.m_identifier_token = decl->m_identifier_token;
-		typed_decl.m_surrounding_function = typed_function.get();
-		typed_function->m_args.push_back(std::move(typed_decl));
+		auto typed_decl = alloc.make<Declaration>();
```
Comment (SebastianMestre) [inline src/typed_ast.cpp:66]: Why did you change this one from a value to a pointer?

**Review by SebastianMestre — APPROVED** [review]
I only skimmed it (I'm pretty tired, sorry). It looks mostly ok, but I found some places where you changed by-reference iteration to by-value (probably bugs), and some places where you changed some by-value members to by-pointer (probably slower). I don't understand why, and would like to know your reasoning.

File: src/typed_ast.hpp (hunk 2 above)
Comment (lushoBarlett) [reply src/typed_ast.hpp:245]: I figured since they are now allocated elsewhere I can't store them here directly, another approach would be to exclude some of the nodes in the node allocator and have them stored here wherever they are needed.

File: src/typed_ast.cpp (hunk 5 above)
Comment (lushoBarlett) [reply src/typed_ast.cpp:66]: This one, honestly, it was a matter of consistency to have them all in the allocator... I guess there are many places where that could be avoided just by not allocating them in the node allocator. I should rethink that, I also did it in the last pr.

**Comment by lushoBarlett** [comment]
I'll dedicate the last of this pr to reverting some things that you pointed out, mainly avoid storing in the allocator when not necessary. I also noticed some places where we could replace `::AST*` with `::Declaration` (in functions), so let me know if you want me to also do that.

File: src/ct_eval.cpp (hunk 1 above)
Comment (lushoBarlett) [reply src/ct_eval.cpp:30]: This one I completely missed it. Probably a misclick in vim.

File: src/typed_ast.cpp (hunk 5 above)
Comment (SebastianMestre) [reply src/typed_ast.cpp:66]: As a rule of thumb, the less indirection the better :)

**Comment by SebastianMestre** [comment]
> I'll dedicate the last of this pr to reverting some things that you pointed out, mainly avoid storing in the allocator when not necessary. I also noticed some places where we could replace `::AST*` with `::Declaration` (in functions), so let me know if you want me to also do that.

That'd be nice

File: src/ct_eval.cpp [hunk 6]
```diff
@@ -86,7 +86,7 @@ TypedAST::IfElseStatement* ct_eval(
 
 TypedAST::ForStatement* ct_eval(
     TypedAST::ForStatement* ast, TypeChecker& tc, TypedAST::Allocator& alloc) {
-	ast->m_declaration = ct_eval(ast->m_declaration, tc, alloc);
+	ct_eval(&ast->m_declaration, tc, alloc);
```
Comment (SebastianMestre) [inline src/ct_eval.cpp:89]: we need to keep this in mind, in case `ct_eval` starts doing more complicated things.

File: src/desugar.cpp [hunk 7]
```diff
@@ -66,8 +66,10 @@ AST* desugar(ShortFunctionLiteral* ast, Allocator& alloc) {
 	block->m_body.push_back(return_stmt);
 
 	auto func = alloc.make<FunctionLiteral>();
-	for (auto& arg : ast->m_args)
-		func->m_args.push_back(desugar(arg, alloc));
+	func->m_args = std::move(ast->m_args);
+
+	for (auto& arg : func->m_args)
+		desugar(&arg, alloc);
```
Comment (SebastianMestre) [inline src/desugar.cpp:72]: ah yeah, the way it was before was pretty dumb

Comment (SebastianMestre) [reply src/desugar.cpp:72]: this is way nicer

File: src/typed_ast.cpp [hunk 8]
```diff
@@ -59,24 +60,21 @@ TypedAST* convert_ast(AST::DictionaryLiteral* ast, Allocator& alloc) {
...
+		typed_function->m_args.push_back(std::move(typed_decl));
 	}
 
 	typed_function->m_body = convert_ast(ast->m_body, alloc);
 
 	return typed_function;
 }
 
-Declaration* convert_ast(AST::Declaration* ast, Allocator& alloc) {
+TypedAST* convert_ast(AST::Declaration* ast, Allocator& alloc) {
```
Comment (SebastianMestre) [inline src/typed_ast.cpp]: That's weird.. why does this not return a `Declaration*`?

**Review by SebastianMestre — APPROVED** [review]
Looks good

File: src/typed_ast.cpp (hunk 8 above)
Comment (lushoBarlett) [reply src/typed_ast.cpp]: I reckon there was a type error. I supposed that it was because I was changing the return type. I find it weird too because I was sure it was like that before.
